### ISSUE TITLE
Z21 work around packet loss

### ISSUE THREAD

**Issue by gfgit** [traintastic/traintastic#60]
Z21 uses UDP protocol, on top of this some implementations (mainly Arduino based one) could be not enough performant to keep up
the message stream and some packets can be lost.
Now imagine triggering STOP because you see some obstacle on the tracks but the packet gets lost...

I suggest we implement some sort of queue of replies waiting to be received.
Each message type will have a sort of priority, higher priority means shorter timeout before sending again and higher retry count.
For example:
```
- Send LAN_X_SET_STOP (high priority)
- Wait 0.5 seconds for `LAN_X_BC_STOPPED`
- If not received retry 5 times.
```
Or:
```
- Send LAN_X_SET_LOCO_FUNCTION (low priority)
- Wait 1 second for `LAN_X_LOCO_INFO`
- If not received retry 2 times.
```

While waiting, other message can be sent/received because protocol is explicitly asynchronous.
Some received messages could look like our reply but are instead generated by other handset controllers connected to Z21.
So we need some logic to track messages, remove them from reply queue when we get a match and schedule retry when timeout expires.

Further optimization would be to use a single timer for all this logic and rework keep alive timer to only trigger when there has not been traffic since n seconds (configurable) but to not trigger when there is already a lot of traffic.

**Comment by gfgit** [comment]
We could show some sort of error message to the user when Z21 seems not to receive our messages

**Comment by reinder** [comment]
Yeah I like this, good suggestion! With UDP packet loss/drop is something to take into account indeed.

> While waiting, other message can be sent/received because protocol is explicitly asynchronous.

I think we should limit the number of messages "in flight" (advanced setting), in case of packet drops we don't want to flood it so more packets are dropped.

> Further optimization would be to use a single timer for all this logic and rework keep alive timer to only trigger when there has not been traffic since n seconds (configurable) but to not trigger when there is already a lot of traffic.

That would be nice, the Traintastic DIY does that too.

p.s. I've contacted the Z21 manufacturer to ask if the will support TCP (in the future), the sadly don't :( for computer control TCP would be much better. (we could still add it, maybe someday someone will implement it)

p.s.2. *advanced settings* in the future I think it would be good to label some settings as advanced, they can be hidden by default in the client, to make it easier for new users.

**Comment by gfgit** [comment]
Well I've discovered Z21 protocol specification allows more messages to be packed in a single UDP datagram.
Z21ServerEmulator implementation was only reading the first one (to be precise first 15 bytes so maybe even dropped on purpose) and then discarded the rest causing all sorts of communication errors.
My code was an adaption of existing Arduino Z21 sketches so probably those do not support this feature either.
I've fixed this issue and discovered another infinite "Async loop" with power state / stop state maybe due to racy startup.

That being said, I think this "Reply queue" is still a valid feature

**Comment by reinder** [comment]
> Well I've discovered Z21 protocol specification allows more messages to be packed in a single UDP datagram. Z21ServerEmulator implementation was only reading the first one (to be precise first 15 bytes so maybe even dropped on purpose) and then discarded the rest causing all sorts of communication errors. My code was an adaption of existing Arduino Z21 sketches so probably those do not support this feature either. I've fixed this issue and discovered another infinite "Async loop" with power state / stop state maybe due to racy startup.

Yes that's correct, Traintastic can do that, especially when many messages are sent in a short time.
It might be good to add a setting for that to put only one message per UDP datagram, so Traintastic can also be used with Z21 that aren't compatible.

LocoNet and XpressNet also have some settings to influence the behavior of the kernel to handle incompatibilities between different implementations. I prefer to add a separate setting for each *fix* and a *command station* setting which acts as a preset with a custom option, e.g. see: [here](https://github.com/traintastic/traintastic/blob/master/server/src/hardware/protocol/loconet/settings.cpp#L139) and [here](https://github.com/traintastic/traintastic/blob/master/server/src/hardware/protocol/xpressnet/settings.cpp#L72), users then can easily select their command station, and all should be correct :) (or use the custom option for something that is not in the list)

> That being said, I think this "Reply queue" is still a valid feature

I agree, that will make it more reliable.

**Comment by gfgit** [comment]
Let's discuss details a bit. I think this reply queue should live inside `ClientKernel` because server doesn't need it.

- Should we store the full sent message in the queue along send time and retry count?
- Should we just store message type instead? But `LanXSetLocoDrive` are specific to decoder address so they should be treated as "per decoder"

I think for some messages we should store whole info, like for `LanXSetLocoDrive` we need loco address and speed step to match received `LanXLocoInfo`.
While simple messages like `LanXSetStop` do not have any particular state.
To keep first implementation simple enough I would store the whole message in a byte vector.

- Should timeout and retry count be fixed for each message type or custom?

I would set an hierarchy of urgent messages and lower priority messages and deduce values from it.

- How do we match request and reply? Like compare loco address of received `LanXLocoInfo` with sent `LanXSetLocoDrive`
Compare turnout address etc...

This is a bit tricky because each message has a different structure so we need custom comparison.

- Should we put a max number of pending requests to avoid running out of memory? What do we do when we reach it?
